src/bin/compress.rs

```diff
@@ -76,43 +79,58 @@ fn main() {
     // assert!(!out_dir_p.exists());
     // std::fs::create_dir(out_dir_p).unwrap();
     
-    let (mut programs, tasks, num_prior_inventions) = args.fmt.load_programs_and_tasks(&args.file).unwrap();
+    let (mut train_programs,
+        test_programs,
+        tasks,
+        num_prior_inventions) = args.fmt.load(&args.file, args.split_train_test).unwrap();
     
     if args.shuffle {
-        programs.shuffle(&mut rand::thread_rng());
+        train_programs.shuffle(&mut rand::thread_rng());
     }
     if let Some(n) = args.truncate {
-        programs.truncate(n);
+        train_programs.truncate(n);
     }
     
     // parse the program strings into expressions
-    let programs: Vec<Expr> = programs.iter().map(|p| p.parse().unwrap()).collect();
+    let train_programs: Vec<Expr> = train_programs.iter().map(|p| p.parse().unwrap()).collect();
+    let test_programs: Option<Vec<Expr>> = if let Some(ps) = test_programs { Some(ps.iter().map(|p| p.parse().unwrap()).collect()) } else { None };
```

Comment:
This "if test data is not None do something with it, otherwise just set it to None" pattern pops up _really_ often in this commit. It's very verbose, which is frustrating because there's almost a much more elegant solution: [Option::map_or](https://doc.rust-lang.org/std/option/enum.Option.html#method.map_or). Unfortunately `map_or` consumes the value, which means you would need to do a bunch of (potentially expensive) `clone`s everywhere. There's an `inspect_or` function that returns a reference instead, but the docs say that it's not on `stable` yet. Which is a little bit annoying...

I might have to write some custom utility function just to make this a bit less ugly, given that it pops up like 10x times...

Reply:
Ah I see. I wouldn't worry too much about the cloning cost as I don't think these are used much in the tight inner loop (theyre more during json loading / writing) and since the compiler might optimize the clone away anyways.

Also in this case where you're reassigning test_programs you're destroying the original value anyways so there's no harm in consuming it / no need for a clone from what I can see, but I see that in other places in the PR thats not true.

I'd use a mix of `map` and `map_or`; in this case you'd just want `map` (so that None stays as None) whereas in some of the later cases you'll want `map_or` with a default of the json `Value::Null`

`map` should be pretty clean.
`if let Some(ps) = test_programs { Some(ps.iter().map(|p| p.parse().unwrap()).collect()) } else { None };`
becomes something like
`test_programs.map(|ps| ps.iter().map(|p| p.parse().unwrap()).collect());`